### ISSUE TITLE
[Bug]: 显卡进程混乱

### ISSUE THREAD

**Issue by wdwz2000** [RLinf/RLinf#259]
### Affected component

进程混乱无法按照指定显卡进行内存分配。
我尝试为进程（libero_10_grpo_openpi.yaml）分配1-3卡使用，但是本应出现在2卡的worker子节点出现在了0卡，即使在run_embodiment.sh中我已经写了export NVIDIA_DRIVER_CAPABILITIES="all"
export CUDA_VISIBLE_DEVICES=1,2,3

### Bug description

我尝试为进程（libero_10_grpo_openpi.yaml）分配1-3卡使用，但是本应出现在2卡的worker子节点出现在了0卡，即使在run_embodiment.sh中我已经写了export NVIDIA_DRIVER_CAPABILITIES="all"
export CUDA_VISIBLE_DEVICES=1,2,3

### Configuration YAML file

YAML config file:libero_10_grpo_openpi.yaml

You can also paste the full config here.
```
defaults:
  - env/train: libero_10
  - env/eval: libero_10
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor,env,rollout: 1-3

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openpi_GRPO"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  val_check_interval: -1
  save_interval: 40
  seq_length: 4096
  max_prompt_length: 50

algorithm:
  auto_reset: True  # 必须true
  ignore_terminations: False
  use_fixed_reset_state_ids: True
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 8
  max_episode_steps: 480
  use_ordered_reset_state_ids: False
  reward_coef: 1.0
  n_chunk_steps: 48
  n_eval_chunk_steps: 48
  num_group_envs: 8
  rollout_epoch: 8
  reward_type: chunk_level
  logprob_type: chunk_level
  entropy_type: token_level

  update_epoch: 4
  adv_type: grpo
  loss_type: actor
  loss_agg_func: "token-mean" 
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  filter_rewards: False
  rewards_lower_bound: 0.1
  rewards_upper_bound: 0.9
  # params for generation
  sampling_params:
    use_greedy: False
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  generation_backend: "huggingface"
  model_dir: "/workspace/openpi/RLinf-Pi0-SFT-Long"
  precision: ${actor.model.precision}
  recompute_logprobs: False
  unnorm_key: libero_10
  enable_offload: False
  pipeline_stage_num: 1

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  checkpoint_load_path: "/workspace/openpi/RLinf-Pi0-SFT-Long"
  checkpoint_save_path: "../results"
  micro_batch_size: 128
  global_batch_size: 2048 # maximum : global batch size = micro batch size * actor num_processes.
  seed: 42
  enable_offload: False

  model:
    model_name: "openpi"
    precision: null
    num_action_chunks: 10 # interface for the env
    action_dim: 7
    is_lora: False
    lora_rank: 32
    gradient_checkpointing: False
    use_wrist_image: True
    use_proprio: True
    num_steps: 5
    # openpi specific parameters
    openpi:
      noise_level: 0.5
      action_chunk: ${actor.model.num_action_chunks}
      num_steps: ${actor.model.num_steps}
      train_expert_only: True
      action_env_dim: ${actor.model.action_dim}
      noise_method: "flow_sde" 
      add_value_head: False
      detach_critic_input: False

  optim:
    lr: 5.0e-6
    value_lr: 1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-05
    clip_grad: 2.0

  fsdp_config:                  # below are fsdp configs, for more details, see https://pytorch.org/docs/stable/fsdp.html and https://huggingface.co/docs/accelerate/package_reference/fsdp

    strategy: "fsdp"            # FSDP strategy: ["fsdp", "fsdp2"]

    # Sharding strategy: "full_shard" (shard parameters, gradients, optimizer states),
    # "shard_grad_op" (shard gradients and optimizer states only), 
    # "hybrid_shard" (combines data parallelism and model parallelism - sharding within FSDP groups, replicate across DDP groups),
    # "no_shard" (no sharding)
    sharding_strategy: "no_shard"

    cpu_offload: False           # whether to offload parameters and gradients to CPU when not in use，if True, actor's enable_offload should be True too.
    offload_pin_memory: False    # whether FSDP2's CPU offload policy should pin memory when cpu_offload is True
    reshard_after_forward: True  # if True, FSDP2 will reshard parameters after forward pass to save memory

    enable_gradient_accumulation: True  # if True, gradient accumulation will be enabled for FSDP/FSDP2 training, which will enhance training performance but increase memory usage.
    forward_prefetch: False             # if True, FSDP will explicitly prefetches the next upcoming all-gather while executing in the forward pass. only use with static graphs. Overlaps communication with computation to improve performance.
    limit_all_gathers: False            # if True, FSDP will synchronizes CPU threads to limit the number of concurrent all-gathers. Only affects strategies that schedule all-gathers
    backward_prefetch: null             # options are null, 'pre', 'post'. if 'pre', FSDP will prefetch the next upcoming all-gather while computing gradients. if 'post', FSDP will prefetch the next all-gather until current gradient is computed.
    use_orig_params: False              # if True, FSDP will use module's original parameters, it means it will expose nn.Module.named_parameters rather than FlatParameter
    use_liger_kernel: False             # if True, liger_kernel will be used for FSDP, note that currently supported models in RLinf are [qwen2.5, qwen2.5-vl], for more details, see liger_kernel's doc

    fsdp_size: -1               # Number of GPUs per FSDP group for hybrid sharding. -1 means use all available GPUs in a single FSDP group

    mixed_precision:            # mixed precision settings for fsdp/fsdp2
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

    amp:
      enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
      precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
      use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.

reward:
  use_reward_model: False

critic:
  use_critic_model: False

```

### Log file

You can find the log file in `logs/` folder or the $output_dir/$experiment_name folder (defined in the yaml config) if you are using our example scripts.

Log file:

If you cannot find the log, please provide the full log messages here.
```
<paste log here>
```

### Environment

python -V:
uv pip list:
nvidia-smi:
RLinf version:
Docker image tag: 

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by andylin-hao** [comment]
Hi @wdwz2000, how did you know the workers are assigned to GPU 0? Are you observing high GPU utilization on device 0?

**Comment by wdwz2000** [comment]
> 你好[@wdwz2000](https://github.com/wdwz2000)你是如何知道工作进程被分配到 GPU 0 的？你是否观察到设备 0 上的 GPU 利用率很高？

我用nvitop查看显卡进程，刚刚又做了个实验，这次我使用显卡2-3，此时显卡2和3上只有进程ray：：EmbodiedFSDPActor.recv_rollout
_batch进程以及ray：：MultiStepRolloutWorker.generate; 而其他本应该出现在显卡2和3上的ray：：EnvWorker.init_worker全部在显卡0和显卡1上

**Comment by wdwz2000** [comment]
> 你好[@wdwz2000](https://github.com/wdwz2000)你是如何知道工作进程被分配到 GPU 0 的？你是否观察到设备 0 上的 GPU 利用率很高？

（注：本服务器已经在跑一个rlinf了，第一个进程在使用libero_10_ppo_openpi.yaml进行测试，分配了4-7卡，这个进程显卡分配没问题；此时bug是在我尝试在同一台服务器的2-3卡再测试一个进程libero_10_grpo_openpi.yaml时发生的）

**Comment by andylin-hao** [comment]
您方便给一下log看看吗，log开头有一些信息可以帮助debug

**Comment by andylin-hao** [comment]
建议去掉CUDA_VISIBLE_DEVICES变量，device的分配是RLinf负责的，设置CUDA_VISIBLE_DEVICES会导致ray看到device和ID不一样。我怀疑是第一个RLinf进程启动时候设置了CUDA_VISIBLE_DEVICES，然后这个RLinf启动了ray于是把GPU限制在了某几个卡上，第二个RLinf启动时候会connect到这个ray上，相关GPU分配受到了影响

**Comment by wdwz2000** [comment]
> 您方便给一下log看看吗，log开头有一些信息可以帮助debug

[gpu_status.txt](https://github.com/user-attachments/files/23493556/gpu_status.txt)

**Comment by wdwz2000** [comment]
> 建议去掉CUDA_VISIBLE_DEVICES变量，device的分配是RLinf负责的，设置CUDA_VISIBLE_DEVICES会导致ray看到device和ID不一样。我怀疑是第一个RLinf进程启动时候设置了CUDA_VISIBLE_DEVICES，然后这个RLinf启动了ray于是把GPU限制在了某几个卡上，第二个RLinf启动时候会connect到这个ray上，相关GPU分配受到了影响

第一个进程启动时，我只设置了export NVIDIA_DRIVER_CAPABILITIES="all"，没有设置CUDA_VISIBLE_DEVICES；我是发现第二个进程出现这个bug后才尝试添加CUDA_VISIBLE_DEVICES限制gpu的

**Comment by andylin-hao** [comment]
> > 您方便给一下log看看吗，log开头有一些信息可以帮助debug
> 
> [gpu_status.txt](https://github.com/user-attachments/files/23493556/gpu_status.txt)

您好，我指的是RLinf运行的log，谢谢

**Comment by wdwz2000** [comment]
> > > 您方便给一下log看看吗，log开头有一些信息可以帮助debug
> > 
> > 
> > [gpu_status.txt](https://github.com/user-attachments/files/23493556/gpu_status.txt)
> 
> 您好，我指的是RLinf运行的log，谢谢

抱歉，发错了

[log_2025.11.12.15.06.txt](https://github.com/user-attachments/files/23493696/log_2025.11.12.15.06.txt)

**Comment by andylin-hao** [comment]
> > > > 您方便给一下log看看吗，log开头有一些信息可以帮助debug
> > > 
> > > 
> > > [gpu_status.txt](https://github.com/user-attachments/files/23493556/gpu_status.txt)
> > 
> > 
> > 您好，我指的是RLinf运行的log，谢谢
> 
> 抱歉，发错了
> 
> [log_2025.11.12.15.06.txt](https://github.com/user-attachments/files/23493696/log_2025.11.12.15.06.txt)

谢谢，请问一下这个log和上面的nvitop是同时获取的吗？我看到这个log里的num_group_envs是2，卡分配的2-3（总数2个），group_size是4，pipeline_stage_num是1，这样每张卡上env总数应该是num_group_envs / num_gpus * group_size * pipeline_stage_num = 2 / 2 * 4 * 1 = 4，但是上面nvitop显示卡0和卡1有8个env进程

**Comment by wdwz2000** [comment]
> > > > > 您方便给一下log看看吗，log开头有一些信息可以帮助debug
> > > > 
> > > > 
> > > > [gpu_status.txt](https://github.com/user-attachments/files/23493556/gpu_status.txt)
> > > 
> > > 
> > > 您好，我指的是RLinf运行的log，谢谢
> > 
> > 
> > 抱歉，发错了
> > [log_2025.11.12.15.06.txt](https://github.com/user-attachments/files/23493696/log_2025.11.12.15.06.txt)
> 
> 谢谢，请问一下这个log和上面的nvitop是同时获取的吗？我看到这个log里的num_group_envs是2，卡分配的2-3（总数2个），group_size是4，pipeline_stage_num是1，这样每张卡上env总数应该是num_group_envs / num_gpus * group_size * pipeline_stage_num = 2 / 2 * 4 * 1 = 4，但是上面nvitop显示卡0和卡1有8个env进程

是同时获取的，我也想知道为什么😂；
此外，这个grpo训不起来，loss_mask有问题Exception occurred while running EmbodiedFSDPActor's function compute_advantages_and_returns: exception is ray::EmbodiedFSDPActor.compute_advantages_and_returns() (pid=4174095, ip=10.90.224.150, actor_id=86e5b130c757ccf2e786257517000000, repr=<rlinf.workers.actor.fsdp_actor_worker.EmbodiedFSDPActor object at 0x7f2e8e328850>)
  File "/opt/venv/.python/cpython-3.11.10-linux-x86_64-gnu/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/.python/cpython-3.11.10-linux-x86_64-gnu/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
           ^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/RLinf/rlinf/scheduler/worker/worker.py", line 70, in sync_func
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/RLinf/rlinf/workers/actor/fsdp_actor_worker.py", line 724, in compute_advantages_and_returns
    advantages_and_returns = calculate_adv_and_returns(**kwargs)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/RLinf/rlinf/algorithms/registry.py", line 109, in calculate_adv_and_returns
    advantages, returns = fn(**kwargs)
                          ^^^^^^^^^^^^
  File "/workspace/RLinf/rlinf/algorithms/registry.py", line 39, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/workspace/RLinf/rlinf/algorithms/advantages.py", line 119, in compute_grpo_advantages
    advantages = (torch.zeros_like(loss_mask) + advantages.view(1, -1)) * loss_mask
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by duguodong7** [comment]
rlinf强制使用了所有的显卡，不知道该怎么指定一部分显卡，比如，CUDA_VISIBLE_DEVICES=2,3,4,5 并不能生效
